api/kernel/fiber.hpp

```diff
@@ -19,45 +19,134 @@
 #ifndef KERNEL_CONTEXT_HPP
 #define KERNEL_CONTEXT_HPP
 
+#include <cstdio>
 #include <delegate>
 
-class Fiber_error : public std::runtime_error {
+class Fiber;
+extern "C" void fiber_jumpstarter(Fiber* f);
+
+/** Exception: General error for fibers */
+class Err_bad_fiber : public std::runtime_error {
+  using runtime_error::runtime_error;
+};
+
+/** Exception: Trying to fetch an object of wrong type  **/
+struct Err_bad_cast : public std::runtime_error {
   using runtime_error::runtime_error;
 };
 
+
 class Fiber {
 
 public:
 
-  using init_func = delegate<void()>;
+  using R_t = void*;
+  using P_t = void*;
+  using init_func = void*(*)(void*);
   using Stack_ptr = std::unique_ptr<char[]>;
   static constexpr int default_stack_size = 0x10000;
   static constexpr int redzone = 0;
 
   void start();
   void restart();
 
-  Fiber(Fiber* parent, int stack_size, init_func func)
+  template<typename R, typename P>
+  Fiber(Fiber* parent, int stack_size, R(*func)(P), void* arg)
```

Comment:
Okay, I understand the rationale now... :+1:
Should probably change the type from `int` to `size_t` to constrain the valid values to natural numbers...